embassy-boot/boot/src/lib.rs

```diff
@@ -1097,99 +1095,48 @@ impl FirmwareUpdater {
 
         trace!("Erased from {} to {}", self.dfu.from, self.dfu.to);
 
-        Ok(FirmwareWriter(self.dfu))
+        Ok(FirmwareWriter::new(self.dfu))
     }
 }
 
 /// FirmwareWriter allows writing blocks to an already erased flash.
-pub struct FirmwareWriter(Partition);
+pub struct FirmwareWriter {
+    partition: Partition,
+    pos: usize,
+}
 
 impl FirmwareWriter {
+    /// Create a new firmware writer.
+    fn new(partition: Partition) -> Self {
+        Self { partition, pos: 0 }
+    }
+
+    /// Get the number of written bytes
+    pub fn pos(&self) -> usize {
+        self.pos
+    }
+
     /// Write data to a flash page.
     ///
     /// The buffer must follow alignment requirements of the target flash and a multiple of page size big.
-    ///
-    /// # Safety
-    ///
-    /// Failing to meet alignment and size requirements may result in a panic.
-    pub async fn write_block<F: AsyncNorFlash>(
-        &mut self,
-        offset: usize,
-        data: &[u8],
-        flash: &mut F,
-        block_size: usize,
-    ) -> Result<(), F::Error> {
-        trace!(
-            "Writing firmware at offset 0x{:x} len {}",
-            self.0.from + offset,
-            data.len()
-        );
-
-        let mut write_offset = self.0.from + offset;
-        for chunk in data.chunks(block_size) {
-            trace!("Wrote chunk at {}: {:?}", write_offset, chunk);
-            flash.write(write_offset as u32, chunk).await?;
-            write_offset += chunk.len();
-        }
-        /*
-        trace!("Wrote data, reading back for verification");
-
```

Comment:
FWIW, the reason for this chunking is that some external flashes cannot write the full page size, so it need to be chunked.

Reply:
As far as I can tell that is a violation of the NorFlash trait then. It does not limit the number of blocks to write at once. I wonder if the block size for those devices could be included in their implementation of the trait?

Reply:
Hmm, maybe. But the flash implementation is a generic 'Qspi' implementation like this https://github.com/embassy-rs/embassy/blob/master/embassy-nrf/src/qspi.rs#L571 where the 'block size' is not known by Qspi, and I'm not sure if  maybe that would be too specific to put in there.

Reply:
Why did you close the PR btw?